include/deal.II/matrix_free/fe_remote_evaluation.h

```diff
@@ -0,0 +1,768 @@
+   * PrecomputedFEEvaluationData. In most cases a one level CRS structure
+   * is enough. In this case @c ptrs has to be constructed and the shift can be
+   * obtained with `get_shift(index)`. The field @c ptrs_ptrs stays empty.
+   * It is only filled if a two level structure is needed. In this case
+   * `get_shift(cell_index, face_number)` return the correct shift.
+   */
+  struct PrecomputedFEEvaluationDataView
+  {
+    /**
+     * Get a pointer to data at index.
+     */
+    unsigned int
+    get_shift(const unsigned int index) const
+    {
+      Assert(ptrs_ptrs.size() == 0, ExcMessage("Two level CRS set up"));
+
+      Assert(index != numbers::invalid_unsigned_int,
+             ExcMessage("Index has to be valid!"));
+
+      Assert(start <= index, ExcInternalError());
+      AssertIndexRange(index - start, ptrs.size());
+      return ptrs[index - start];
+    }
+
+    /**
+     * Get a pointer to data at, e.g. (cell_index, face_number).
+     */
+    unsigned int
+    get_shift(const unsigned int cell_index,
+              const unsigned int face_number) const
+    {
+      Assert(ptrs_ptrs.size() > 0, ExcMessage("No two level CRS set up"));
+
+      Assert(cell_index != numbers::invalid_unsigned_int,
+             ExcMessage("Cell index has to be valid!"));
+      Assert(face_number != numbers::invalid_unsigned_int,
+             ExcMessage("Face number has to be valid!"));
+
+      Assert(start <= cell_index, ExcInternalError());
+
+      AssertIndexRange(cell_index - start, ptrs_ptrs.size());
+      const unsigned int face_index =
+        ptrs_ptrs[cell_index - start] + face_number;
+      AssertIndexRange(face_index, ptrs.size());
+      return ptrs[face_index];
+    }
+
+    /**
+     * Get the number of stored values.
+     */
+    unsigned int
+    size() const
+    {
+      Assert(ptrs.size() > 0, ExcInternalError());
+      return ptrs.back();
+    }
+
+    /**
+     * This parameter can be used if indices do not start with 0.
+     */
+    unsigned int start = 0;
+
+    /**
+     * Pointers to @c ptrs.
+     */
+    std::vector<unsigned int> ptrs_ptrs;
+
+    /**
+     * Pointers to data.
+     */
+    std::vector<unsigned int> ptrs;
+  };
+
+  /**
+   * A class helps to access @c PrecomputedFEEvaluationData in a thread safe
+   * manner. Each thread has to create this wrapper class on its own to
+   * avoid race-conditions during @c reinit().
+   */
+  template <int dim, int n_components, typename value_type>
+  class PrecomputedFEEvaluationDataAccessor
+  {
+  public:
+    PrecomputedFEEvaluationDataAccessor(
+      const PrecomputedFEEvaluationData<dim, n_components, value_type> &data,
+      const PrecomputedFEEvaluationDataView                            &view)
+      : view(view)
+      , data(data)
+      , data_offset(numbers::invalid_unsigned_int)
+    {}
+
+    /**
+     * Get the value at quadrature point @p q. The entity on which the values
+     * are defined is set via `reinit()`.
+     */
+    const auto
```

Comment:
We need to be explicit regarding the type.

Reply:
[d364063](https://github.com/dealii/dealii/pull/16300/commits/d3640637ccd99b4be404153d92cc1b1d3d0fa841)